scanner/src/main/kotlin/experimental/ScannerWrapper.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2021 HERE Europe B.V.
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.File
+
+import org.ossreviewtoolkit.model.KnownProvenance
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.ScanSummary
+import org.ossreviewtoolkit.model.ScannerDetails
+import org.ossreviewtoolkit.model.config.ScannerConfiguration
+import org.ossreviewtoolkit.scanner.ScanResultsStorage
+import org.ossreviewtoolkit.scanner.ScannerCriteria
+
+/**
+ * An interface to wrap source code scanners.
+ */
+sealed interface ScannerWrapper {
```

Comment:
The `sealed` in here prevents us from adding scanner implementations via plugins as [all implementations of a sealed interface are known at compile time](https://kotlinlang.org/docs/whatsnew15.html#sealed-interfaces). That's a problem as we have concrete plans to add scanner plugins.

Reply:
Fortunately that's not an issue with a `sealed interface`. In this case you cannot make a class or interface inherit directly from `ScannerWrapper` in another module, but it is allowed to have classes or interfaces inherit from one of the sub-interfaces (like `LocalScannerWrapper`) in another module.

Reply:
> but it is allowed to have classes or interfaces inherit from one of the sub-interfaces (like `LocalScannerWrapper`) in another module.

Interesting, do you have a reference for that? Or did you actually try it out?

Reply:
See: https://kotlinlang.org/docs/sealed-classes.html#location-of-direct-subclasses

> Direct subclasses of sealed classes and interfaces must be declared in the same package. [...] These restrictions don't apply to indirect subclasses. If a direct subclass of a sealed class is not marked as sealed, it can be extended in any ways that its modifiers allow.

This makes this a good use case for a `sealed interface`. We ensure that no new kind of scanner can be added by a plugin, because ORT would not know how to handle it, but new implementations of existing kinds can be added. However, while writing this I realized that `interface RemoteScannerWrapper : ScannerWrapper` also needs to be sealed, I will update the PR.